File: contracts/api/clients/jakarta/src/test/java/app/epistola/client/jakarta/validation/schema/TemplateSchemaValidatorTest.java
// SPDX-FileCopyrightText: Epistola Nederland B.V.
//
// SPDX-License-Identifier: EUPL-1.2

package app.epistola.client.jakarta.validation.schema;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.junit.jupiter.api.Assertions.assertTrue;

import app.epistola.client.jakarta.FakeApis;
import app.epistola.client.jakarta.api.TemplatesApi;
import app.epistola.client.jakarta.model.TemplateDto;
import java.time.Duration;
import java.util.List;
import java.util.Map;
import java.util.concurrent.atomic.AtomicInteger;
import org.junit.jupiter.api.Test;

class TemplateSchemaValidatorTest {

    private static final Map<String, Object> INVOICE_SCHEMA = Map.of(
            "$schema", "https://json-schema.org/draft/2020-12/schema",
            "type", "object",
            "required", List.of("customerName", "total"),
            "properties",
            Map.of(
                    "customerName", Map.of("type", "string", "minLength", 1),
                    "total", Map.of("type", "number", "minimum", 0)));

    @Test
    void data_that_satisfies_the_schema_passes() {
        TemplateSchemaValidator validator = new TemplateSchemaValidator(new StubTemplatesApi(INVOICE_SCHEMA).api());

        validator.validate("acme-corp", "default", "invoice", Map.of("customerName", "Jane", "total", 42.5));
    }

    @Test
    void a_missing_required_field_is_reported_with_its_path_and_keyword() {
        TemplateSchemaValidator validator = new TemplateSchemaValidator(new StubTemplatesApi(INVOICE_SCHEMA).api());

        TemplateDataValidationException e = assertThrows(
                TemplateDataValidationException.class,
                () -> validator.validate("acme-corp", "default", "invoice", Map.of("total", 42.5)));

        assertEquals(1, e.getErrors().size());
        assertEquals("required", e.getErrors().get(0).getKeyword());
        assertTrue(e.getMessage().contains("1 error"), e.getMessage());
        assertTrue(e.formatErrors().contains("customerName"), e.formatErrors());
    }

    @Test
    void every_violation_is_collected_rather_than_just_the_first() {
        TemplateSchemaValidator validator = new TemplateSchemaValidator(new StubTemplatesApi(INVOICE_SCHEMA).api());

        TemplateDataValidationException e = assertThrows(
                TemplateDataValidationException.class,
                () -> validator.validate(
                        "acme-corp", "default", "invoice", Map.of("customerName", "", "total", -1)));

        assertEquals(2, e.getErrors().size(), e.formatErrors());
    }

    @Test
    void a_template_without_a_schema_validates_anything() {
        TemplateSchemaValidator validator = new TemplateSchemaValidator(new StubTemplatesApi(null).api());

        validator.validate("acme-corp", "default", "invoice", Map.of("whatever", true));
    }

    @Test
    void a_draft_07_schema_is_compiled_under_draft_07_rules() {
        Map<String, Object> draft07 = Map.of(
                "$schema", "http://json-schema.org/draft-07/schema#",
                "type", "object",
                "required", List.of("id"),
                "properties", Map.of("id", Map.of("type", "string")));
        TemplateSchemaValidator validator = new TemplateSchemaValidator(new StubTemplatesApi(draft07).api());

        validator.validate("acme-corp", "default", "invoice", Map.of("id", "x"));
        assertThrows(
                TemplateDataValidationException.class,
                () -> validator.validate("acme-corp", "default", "invoice", Map.of("id", 1)));
    }

    @Test
    void a_schema_with_no_dollar_schema_still_compiles() {
        Map<String, Object> bare = Map.of("type", "object", "required", List.of("id"));
        TemplateSchemaValidator validator = new TemplateSchemaValidator(new StubTemplatesApi(bare).api());

        assertThrows(
                TemplateDataValidationException.class,
                () -> validator.validate("acme-corp", "default", "invoice", Map.of()));
    }

    @Test
    void the_template_is_fetched_once_and_then_served_from_the_cache() {
        StubTemplatesApi api = new StubTemplatesApi(INVOICE_SCHEMA);
        TemplateSchemaValidator validator = new TemplateSchemaValidator(api.api());

        for (int i = 0; i < 5; i++) {
            validator.validate("acme-corp", "default", "invoice", Map.of("customerName", "Jane", "total", 1));
        }

        assertEquals(1, api.fetches.get(), "a schema fetch per generation call would double the round trips");
    }

    @Test
    void evicting_the_cache_forces_a_refetch() {
        StubTemplatesApi api = new StubTemplatesApi(INVOICE_SCHEMA);
        TtlSchemaCache cache = new TtlSchemaCache(Duration.ofMinutes(5));
        TemplateSchemaValidator validator = new TemplateSchemaValidator(api.api(), cache);

        validator.validate("acme-corp", "default", "invoice", Map.of("customerName", "Jane", "total", 1));
        cache.evict("acme-corp", "default", "invoice");
        validator.validate("acme-corp", "default", "invoice", Map.of("customerName", "Jane", "total", 1));

        assertEquals(2, api.fetches.get());
    }

    @Test
    void an_expired_entry_is_reloaded() throws Exception {
        StubTemplatesApi api = new StubTemplatesApi(INVOICE_SCHEMA);
        TemplateSchemaValidator validator =
                new TemplateSchemaValidator(api.api(), new TtlSchemaCache(Duration.ofMillis(50)));

        validator.validate("acme-corp", "default", "invoice", Map.of("customerName", "Jane", "total", 1));
        Thread.sleep(80);
        validator.validate("acme-corp", "default", "invoice", Map.of("customerName", "Jane", "total", 1));

        assertEquals(2, api.fetches.get());
    }

    @Test
    void the_same_template_id_in_two_catalogs_is_two_cache_entries() {
        // Two catalogs of one tenant can both hold an "invoice" template, with different schemas.
        // Keying on (tenant, template) alone would validate one against the other's contract.
        StubTemplatesApi api = new StubTemplatesApi(INVOICE_SCHEMA);
        TemplateSchemaValidator validator = new TemplateSchemaValidator(api.api());

        validator.validate("acme-corp", "catalog-a", "invoice", Map.of("customerName", "Jane", "total", 1));
        validator.validate("acme-corp", "catalog-b", "invoice", Map.of("customerName", "Jane", "total", 1));

        assertEquals(2, api.fetches.get());
    }

    @Test
    void different_templates_get_different_cache_entries() {
        StubTemplatesApi api = new StubTemplatesApi(INVOICE_SCHEMA);
        TemplateSchemaValidator validator = new TemplateSchemaValidator(api.api());

        validator.validate("acme-corp", "default", "invoice", Map.of("customerName", "Jane", "total", 1));
        validator.validate("acme-corp", "default", "receipt", Map.of("customerName", "Jane", "total", 1));
        validator.validate("other-corp", "default", "invoice", Map.of("customerName", "Jane", "total", 1));

        assertEquals(3, api.fetches.get());
    }

    @Test
    void a_null_ttl_is_rejected_rather_than_caching_forever() {
        assertThrows(IllegalArgumentException.class, () -> new TtlSchemaCache(Duration.ZERO));
        assertThrows(IllegalArgumentException.class, () -> new TtlSchemaCache(null));
    }

    /** A TemplatesApi that serves one schema and counts fetches, so caching is observable. */
    private static final class StubTemplatesApi {

        private final AtomicInteger fetches = new AtomicInteger();
        private final TemplatesApi api;

        private StubTemplatesApi(Object schema) {
            this.api = FakeApis.of(TemplatesApi.class, Map.of("getTemplate", args -> {
                fetches.incrementAndGet();
                String tenantId = (String) args[0];
                String templateId = (String) args[2];
                return new TemplateDto().id(templateId).tenantId(tenantId).name(templateId).schema(schema);
            }));
        }

        private TemplatesApi api() {
            return api;
        }
    }
}

File: contracts/api/clients/nodejs-fetch/test/validation/templateSchemaValidator.test.ts
// SPDX-FileCopyrightText: Epistola Nederland B.V.
//
// SPDX-License-Identifier: EUPL-1.2

import assert from 'node:assert/strict'
import { test } from 'node:test'
import {
  TemplateDataValidationException,
  TemplateSchemaValidator,
  TtlSchemaCache,
  ValidatingGenerationApi,
  type GenerationApiLike,
  type GenerationJobResponse,
  type TemplateSchemaSource,
} from '../../src/index.js'

const SCHEMA = {
  type: 'object',
  required: ['name'],
  properties: {
    name: { type: 'string' },
    age: { type: 'integer', minimum: 0 },
    email: { type: 'string', format: 'email' },
  },
}

class StubTemplatesApi implements TemplateSchemaSource {
  calls = 0
  constructor(private readonly schema: object | undefined) {}
  async getTemplate(): Promise<{ schema?: object }> {
    this.calls++
    return { schema: this.schema }
  }
}

class StubGenerationApi implements GenerationApiLike {
  generated: unknown[] = []
  private readonly job: GenerationJobResponse = { requestId: '88888888-8888-4888-8888-000000000001', status: 'PENDING', jobType: 'SINGLE', totalCount: 1, createdAt: new Date() }
  async generateDocument(params: unknown): Promise<GenerationJobResponse> {
    this.generated.push(params)
    return this.job
  }
  async generateDocumentBatch(params: unknown): Promise<GenerationJobResponse> {
    this.generated.push(params)
    return this.job
  }
}

test('valid data passes', async () => {
  await new TemplateSchemaValidator(new StubTemplatesApi(SCHEMA)).validate('t', 'c', 'tpl', { name: 'Ada', age: 30 })
})

test('invalid data fails with field-level failures', async () => {
  await assert.rejects(new TemplateSchemaValidator(new StubTemplatesApi(SCHEMA)).validate('t', 'c', 'tpl', { age: -1, email: 'nope' }), (error: unknown) => {
    assert.ok(error instanceof TemplateDataValidationException)
    assert.deepEqual(error.errors.map((f) => [f.path, f.keyword]), [
      ['age', 'minimum'],
      ['email', 'format'],
      ['name', 'required'],
    ])
    assert.match(error.formatErrors(), /name: must have required property/)
    return true
  })
})

test('a template without a schema is a no-op', async () => {
  await new TemplateSchemaValidator(new StubTemplatesApi(undefined)).validate('t', 'c', 'tpl', { whatever: true })
})

test('the schema is cached between calls', async () => {
  const api = new StubTemplatesApi(SCHEMA)
  const validator = new TemplateSchemaValidator(api)
  await validator.validate('t', 'c', 'tpl', { name: 'a' })
  await validator.validate('t', 'c', 'tpl', { name: 'b' })
  assert.equal(api.calls, 1)
})

test('the same template id in two catalogs is two cache entries', async () => {
  // Two catalogs of one tenant can both hold a "tpl" template, with different schemas. Keying on
  // (tenant, template) alone would validate one against the other's contract.
  const api = new StubTemplatesApi(SCHEMA)
  const validator = new TemplateSchemaValidator(api)
  await validator.validate('t', 'catalog-a', 'tpl', { name: 'a' })
  await validator.validate('t', 'catalog-b', 'tpl', { name: 'b' })
  assert.equal(api.calls, 2)
})

test('a TTL cache expires and can be evicted', async () => {
  const api = new StubTemplatesApi(SCHEMA)
  const cache = new TtlSchemaCache(20)
  const validator = new TemplateSchemaValidator(api, cache)
  await validator.validate('t', 'c', 'tpl', { name: 'a' })
  await new Promise((resolve) => setTimeout(resolve, 30))
  await validator.validate('t', 'c', 'tpl', { name: 'a' })
  assert.equal(api.calls, 2)
  cache.evict('t', 'c', 'tpl')
  await validator.validate('t', 'c', 'tpl', { name: 'a' })
  assert.equal(api.calls, 3)
  assert.throws(() => new TtlSchemaCache(0), RangeError)
})

test('a 2020-12 schema is validated with that dialect', async () => {
  const schema = {
    $schema: 'https://json-schema.org/draft/2020-12/schema',
    type: 'object',
    properties: { lines: { type: 'array', prefixItems: [{ type: 'string' }], items: false } },
  }
  const validator = new TemplateSchemaValidator(new StubTemplatesApi(schema))
  await validator.validate('t', 'c', 'tpl', { lines: ['ok'] })
  await assert.rejects(validator.validate('t', 'c', 'tpl', { lines: ['ok', 'extra'] }), TemplateDataValidationException)
})

test('ValidatingGenerationApi validates before delegating', async () => {
  const generation = new StubGenerationApi()
  const api = new ValidatingGenerationApi(generation, new StubTemplatesApi(SCHEMA))
  await assert.rejects(
    api.generateDocument({ tenantId: 't', generateDocumentRequest: { catalogId: 'c', templateId: 'tpl', data: { age: -5 } } }),
    TemplateDataValidationException,
  )
  assert.deepEqual(generation.generated, [])
  await api.generateDocument({ tenantId: 't', generateDocumentRequest: { catalogId: 'c', templateId: 'tpl', data: { name: 'ok' } } })
  assert.equal(generation.generated.length, 1)
})

test('batch validation aggregates errors with the item index', async () => {
  const generation = new StubGenerationApi()
  const api = new ValidatingGenerationApi(generation, new StubTemplatesApi(SCHEMA))
  await assert.rejects(
    api.generateDocumentBatch({
      tenantId: 't',
      generateBatchRequest: {
        items: [
          { catalogId: 'c', templateId: 'tpl', data: { name: 'ok' } },
          { catalogId: 'c', templateId: 'tpl', data: {} },
        ],
      },
    }),
    (error: unknown) => {
      assert.ok(error instanceof TemplateDataValidationException)
      assert.ok(error.errors.some((f) => f.path.startsWith('items[1]')))
      return true
    },
  )
  assert.deepEqual(generation.generated, [])
})
